sdk/core/core/src/az_aad.c

```diff
@@ -113,29 +59,36 @@ AZ_NODISCARD az_result _az_aad_build_body(
     az_span client_secret,
     az_span* out_body)
 {
-  az_span grant_type_and_client_id_key
+  az_span const grant_type_and_client_id_key
       = AZ_SPAN_FROM_STR("grant_type=client_credentials&client_id=");
   AZ_RETURN_IF_NOT_ENOUGH_SIZE(body, az_span_size(grant_type_and_client_id_key));
+
   az_span remainder = az_span_copy(body, grant_type_and_client_id_key);
+  int32_t url_length = 0;
 
-  AZ_RETURN_IF_FAILED(_az_url_encode(client_id, &remainder));
+  AZ_RETURN_IF_FAILED(_az_url_encode(remainder, client_id, &url_length));
+  remainder = az_span_slice_to_end(remainder, url_length);
 
-  az_span scope_key = AZ_SPAN_FROM_STR("&scope=");
+  az_span const scope_key = AZ_SPAN_FROM_STR("&scope=");
   AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, az_span_size(scope_key));
+
   remainder = az_span_copy(remainder, scope_key);
 
-  AZ_RETURN_IF_FAILED(_az_url_encode(scopes, &remainder));
+  AZ_RETURN_IF_FAILED(_az_url_encode(remainder, scopes, &url_length));
+  remainder = az_span_slice_to_end(remainder, url_length);
 
   if (az_span_size(client_secret) > 0)
   {
-    az_span client_secret_key = AZ_SPAN_FROM_STR("&client_secret=");
+    az_span const client_secret_key = AZ_SPAN_FROM_STR("&client_secret=");
     AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, az_span_size(client_secret_key));
+
     remainder = az_span_copy(remainder, client_secret_key);
 
-    AZ_RETURN_IF_FAILED(_az_url_encode(client_secret, &remainder));
+    AZ_RETURN_IF_FAILED(_az_url_encode(remainder, client_secret, &url_length));
+    remainder = az_span_slice_to_end(remainder, url_length);
   }
 
-  *out_body = az_span_slice(body, 0, _az_span_diff(remainder, body));
+  *out_body = az_span_slice(body, 0, url_length);
```

Comment:
Are you sure this `url_length` is correct and the `out_body` contains everything we need? Do we have tests to verify because it doesn't look like things like `"&client_secret="` are accounted for here. If we don't, please add some and all we are doing is returning a small slice of the body from the last `url_length` value.

Reply:
Thank you I actually just debugged and fixed this, fix is coming soon.

Reply:
Done.

Reply:
Add more unit tests to capture bugs like these in the future.

Reply:
Thanks for the suggestion, I'm working on it, if everything plays out well, it it will be a next commit to this PR. Otherwise if it takes too much time, I'll put that to backlog.

Reply:
Done :) New test covers token retrieval, token caching, token refresh, and then token caching once again. It covers client secret credential end -to-end - from initialization to getting a HTTP response via the request with the authentication header set to the expected bearer token value.